lizmap/modules/lizmap/lib/Project/QgisProject.php

```diff
@@ -886,8 +916,39 @@ public function readEditionLayers($editionLayers)
      */
     public function readEditionForms($editionLayers, $proj)
     {
+        $embeddedEditionLayers = array();
         foreach ($editionLayers as $key => $obj) {
-            $layerXml = $this->getXmlLayer2($this->xml, $obj->layerId);
+            $xml = null;
+
+            // check for embedded layers
+            $layerDefinition = $this->getLayerDefinition($obj->layerId);
+            if ($layerDefinition && array_key_exists('embedded', $layerDefinition) && $layerDefinition['embedded'] == 1) {
+                if (array_key_exists($layerDefinition['projectPath'], $embeddedEditionLayers)) {
+                    // use QgisProject instance already created
+                    $qgisProject = $embeddedEditionLayers[$layerDefinition['projectPath']];
+                } else {
+                    // create new QgisProject instance or retreive it from cache, if any
+                    $path = realpath(dirname($this->path).DIRECTORY_SEPARATOR.$layerDefinition['projectPath']);
+                    $qgsMtime = filemtime($path);
+                    $qgsCfgMtime = filemtime($path.'.cfg');
+                    $cacheHandler = new ProjectCache($path, $qgsMtime, $qgsCfgMtime, $this->appContext);
+                    $data = $cacheHandler->retrieveProjectData();
+
+                    if ($data) {
+                        $qgisProject = new QgisProject($path, $this->services, $this->appContext, $data['qgis']);
+                    } else {
+                        $qgisProject = new QgisProject($path, $this->services, $this->appContext);
+                    }
+
+                    $embeddedEditionLayers[$layerDefinition['projectPath']] = $qgisProject;
+                }
+
+                $xml = $qgisProject->getXml();
+            } else {
+                $xml = $this->xml;
+            }
+
+            $layerXml = $this->getXmlLayer2($xml, $obj->layerId);
```

Comment:
I think you should create a new dedicated private function to extract this information and use it in `readEditionLayers` and `readEditionForms` methods. This will avoid to duplicate the code